### PR TITLE
[WIP] [IDE] Toy with reimplementing synthesized extension checking.

### PR DESCRIPTION
**Pull request by DougGregor**
[IDE] Toy with reimplementing synthesized extension checking.  …
The isApplicable() operation for extensions re-implements the idea of
checking whether a particular extension is valid for a given
type. Replace its implementation with one based on
swift::isExtensionApplied(), generalizing the latter to support this
use case.

### CODE REVIEW

File: lib/Sema/CSGen.cpp [hunk 1]
```diff
@@ -3473,6 +3474,21 @@ bool swift::isExtensionApplied(DeclContext &DC, Type BaseTy,
...
+    // Open up the type parameters.
+    constraints::OpenedTypeMap replacements;
+    CS.openGeneric(&DC, DC.getModuleScopeContext(),
+                   DC.getGenericSignatureOfContext(),
+                   /*skipProtocolSelfConstraint=*/false,
+                   Loc, replacements);
+    BaseTy = CS.openType(BaseTy, replacements);
+
+    // Allow free type variables.
+    allowFreeTypeVariables = FreeTypeVariableBinding::Allow;
```
Comment (DougGregor) [inline lib/Sema/CSGen.cpp]: Looks like we aren't using `allowFreeTypeVariables` anywhere, but it should be passed to `solveSingle`

Comment (DougGregor) [reply lib/Sema/CSGen.cpp]: On success, need to map remaining constraints back to Requirements for the caller

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**